[Kappaccinoh] iP

Kappaccinoh opened the pull request:
# Hi there! I am CampusBot!

## I am a **work in process** bot off the original Duke ChatBot, watch me _grow_ into a full fledged ~~human~~ chatbot with cool and amazing features! 😃 👍  

### Created by [Kappaccinoh](https://www.linkedin.com/in/jia-wei-lim-747037181/) for completion of CS2103T

## Usage
Compile using `javac Campus.java` and execute using `java Campus`

## Current Improved Features
- Find Tasks
- Save Tasks to file in text format
- Dates and Times in java dtf-format
  - Was previously just a string (may expand to allowing for both types)
 
## Current Bugs/Improvements needed to be made (in priority)
1. Full Testing Coverage
2. Organise files into packages
3. Need to fix running the main CampusTest
```java
  @Test
  public void campusBotInitialises() {
      // Redirect System.out to capture console output
      ByteArrayOutputStream outContent = new ByteArrayOutputStream();
      System.setOut(new PrintStream(outContent));
  
      // Run the method
      String filePath = "src/test/java/campus/dataTest.txt";
      Campus campusBot = new Campus(filePath);
      campusBot.run();
  
      // Set up simulated input
      String simulatedInput = "bye";
      InputStream originalIn = System.in;
      System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
  
      // Capture and trim the console output
      String consoleOutput = outContent.toString();
  
      // Restore System.out
      System.setOut(System.out);
  
      String expected = "------------------------------\n" +
              "Hello! I'm Campus\n" +
              "What can I do for you?\n" +
              "\n" +
              "------------------------------\n\n";
  
      // Assert the expected output
      assertEquals(expected, consoleOutput);
  }
```
> [!NOTE]
> Not exactly sure how to remove the data file if it exists - needs more work and research

## Week 4 Progress
- [x] ip component
- [x] tp component

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Kappaccinoh committed:
Rename Greet Exit

Kappaccinoh committed:
Echo

Kappaccinoh committed:
Add List

Kappaccinoh committed:
Added task class and modified campus class logic

Kappaccinoh committed:
Finished Mark and Unmark logic

Kappaccinoh committed:
Added 3 new classes which extends the abstract class task

Kappaccinoh committed:
done with class logics and their respective string formattings

Kappaccinoh committed:
Automated Text UI Testing

Kappaccinoh committed:
Added in logic to handle errors and created a new error class

Kappaccinoh committed:
Added Delete Task Functionality:

Kappaccinoh committed:
Ammended task classes and implemented partial data file logic

Kappaccinoh committed:
completed functionality for writing to text files

Kappaccinoh committed:
refactoring code for readability

Kappaccinoh committed:
Big Commit, changed gitignore, refactored strings to use datetimeformat, handled improper datetime formatting with exceptions in individual classes

Kappaccinoh committed:
merge commit with branch-level-7 no ff

Kappaccinoh committed:
merge commit with branch-level-8 no ff

Kappaccinoh committed:
refactored into more classes

Kappaccinoh committed:
created package

Kappaccinoh committed:
merge commit with moreoop branch

Kappaccinoh committed:
merge commit with a-packages

Kappaccinoh committed:
further packaging into sub packages

Kappaccinoh committed:
merge commit with updates packages branch

Kappaccinoh committed:
merging w gradle no ff

Kappaccinoh committed:
added in Junit test cases

Kappaccinoh committed:
merge commit with JUnit branch noff

Kappaccinoh committed:
added more javadocs for longer functions

Kappaccinoh committed:
minor changes to code according to java coding standards

Kappaccinoh committed:
added find functionality

Kappaccinoh committed:
merge commit with javadoc branch

Kappaccinoh committed:
merge commit with coding standard branch

Kappaccinoh committed:
resolved merge conflicts with level-9 find branch

Kappaccinoh committed:
setting up checkstyle but still has issues

Kappaccinoh committed:
added static fxml files - preset up of javafx but after modifying gradle

Kappaccinoh committed:
installed successfully logic of controllers not working refer to handleUserInput in MainWindow.java file for issues

Kappaccinoh committed:
solved crash issue, wrong routing for dialogbox and mainwindow for source files

Kappaccinoh committed:
refactored code for javafx GUI to fit with chatbot logic

Kappaccinoh committed:
removed testing for console driven testing

Kappaccinoh committed:
fixed minor exit bugs and added wrapper for messages

Kappaccinoh committed:
merge commit with Level-10 no ff

Kappaccinoh committed:
checkstyle working pre checkstyle fixes

Kappaccinoh committed:
Added changes which violate the checkstyle

Kappaccinoh committed:
fixing merge conflicts with checkstyle branch and master

LargeCrowd commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,41 @@
+     * Creation of a Campus instance
+     */
+    public Campus() {
+        String filePath = "src/main/java/campus/dataTest.txt";
```
Should this be in SCREAMING_SNAKE_FONT?
```suggestion
        String FILE_PATH= "src/main/java/campus/dataTest.txt";
```

LargeCrowd commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,41 @@
+        return this.parser.respond(input);
+    }
+
+    public String greet() {
+        return this.ui.greet();
+    }
+
+    public String exit() {
+        return this.ui.exit();
+    }
```
Would it be better if the naming of methods are more specific?
```suggestion
    public String printGreeting() {
        return this.ui.printGreeting();
    }

    public String exitProgram() {
        return this.ui.exitProgram();
    }
```

LargeCrowd commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,41 @@
+        new Campus();
+    }
+
+    public String respond(String input) {
+        return this.parser.respond(input);
+    }
```
The naming for this leads to some inclarity regarding its function. Also consider providing javadocs for methods in this class?

LargeCrowd commented on src/main/java/campus/infrastructure/Parser.java:
```diff
@@ -0,0 +1,217 @@
+        }
+    }
+
+    /**
+     * Responds on user input
+     */
+    public String respond(String userInput) {
+        try {
+            return this.processCommand(userInput);
+        } catch (CampusException e) {
+            return this.ui.displayErrorMessage(e);
+        }
+    }
```
Consider providing some clearer naming for this method. Maybe something like "executeUserInput"?

LargeCrowd commented on src/main/java/campus/infrastructure/Storage.java:
```diff
@@ -0,0 +1,75 @@
+     * @return Returns a list of strings, each string contains information about the DB, +
+     *      null if initialised for the first time.
+     */
+    public List<String> readFromDBCreateIfNotExists(String filePath) throws IOException {
```
Naming seems a little bit long. Could consider a more concise naming convention
```suggestion
    public List<String> createTaskList(String filePath) throws IOException {
```

Kappaccinoh replied:
thank you for the suggestion! I was thinking about how to better name this as well but I feel that `createTaskList` doesnt quite capture what the function does, maybe I can change it to another name but throw the behaviour in the documentation.

LargeCrowd commented on src/main/java/campus/infrastructure/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+    public int size() {
+        return tasks.size();
+    }
+    public void delete(Task task) {
```
```suggestion
    public void deleteTask(Task task) {
```

LargeCrowd commented on src/main/java/campus/infrastructure/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+        return this.tasks.get(index);
+    }
+
+    public int size() {
```
```suggestion
    public int getListSize() {
```

LargeCrowd commented on src/main/java/campus/infrastructure/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+        task.markIncomplete();
+    }
+
+    public void add(Task task) {
```
```suggestion
    public void addTask(Task task) {
```

LargeCrowd commented on src/main/java/campus/infrastructure/Ui.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Produces a greeting message
+     */
+    public String greet() {
```
```suggestion
    public String printGreeting() {
```

LargeCrowd commented on src/main/java/campus/infrastructure/Ui.java:
```diff
@@ -0,0 +1,125 @@
+     * Displays all the current items in the TaskList in a specified format with numbering
+     * @param taskList the object containing all the tasks in a list
+     */
+    public String display(TaskList taskList) {
```
```suggestion
    public String printTaskList(TaskList taskList) {
```

LargeCrowd reviewed (commented):
Overall seems well organised and easy to read. Method naming in certain sections could be clearer, and consider adding javadocs for some methods. Good job!

Kappaccinoh committed:
responded to github pull request review comments on variable and function naming changes

Kappaccinoh committed:
merge commit no ff with minor fixes

Kappaccinoh committed:
added some assertions in the infrastructure files

Kappaccinoh committed:
swapped out task list for stream and modified corresponding methods for it

Kappaccinoh committed:
merge commit no ff with assertions branch

Kappaccinoh committed:
resolving merge conflicts

Kappaccinoh committed:
Refactored and Reviewed Code Base

Parsing logic in Parser.java uses a series of case statements to
determine the correct response to a user's input. The function
itself looks very long and untidy.

I extracted out the lines and separated them into different
functions.

This is for readability and also to conceptually divide the
parsing into the types of commands that a user might give the
ChatBot.

I'm happy with the way the files are structured for now but in the
future we might want to organise the files into classes/objects
if the functionality expands and parsing becomes unfeasible or
too complicated.

Kappaccinoh committed:
merge commit for master for codequality branch

Kappaccinoh committed:
adding YML file for github workflows

Kappaccinoh committed:
changing yml file

Kappaccinoh committed:
merge commit with CI github workflows

E0735389 commented on src/main/java/campus/infrastructure/Parser.java:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Handles the commands for mark, unmark and delete - passes it to the taskList object which contains methods
+     * to deal with manipulation of the task
+     * @param command mark/unmark/delete
```
The Java coding standard requires that punctuation should be present. This mistake can also be found in several other parts of the code.

E0735389 commented on src/main/java/campus/data.txt:
```diff
@@ -0,0 +1,3 @@
+T | 0 | do homework
+E | 0 | dinner with friends | 1900 29/01/2024 | 2200 29/01/2024
+D | 0 | buy cny decorations | 2359 30/01/2024
```
I'm not sure if you're supposed to include the data file into the repository as source code. Consider using gitignore.

E0735389 commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,54 @@
+
+    public static void main(String[] args) {
+        new Campus();
+    }
```
Is the description on how to launch the program still accurate? Since this main function looks like it wouldn't do anything.

E0735389 commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @SuppressWarnings("checkstyle:AbbreviationAsWordInName")
+    public Campus() {
+        String FILE_PATH = "src/main/java/campus/dataTest.txt";
```
Consider making this `final`.

E0735389 commented on src/main/java/campus/infrastructure/Ui.java:
```diff
@@ -0,0 +1,132 @@
+     * Displays all the current items in the TaskList in a specified format with numbering
+     * @param taskList the object containing all the tasks in a list
+     */
+    public String printTaskList(TaskList taskList) {
```
This may be a bad name for this method, given that it doesn't actually print it out, just return the output as a string. Maybe `formatTaskList`?

Kappaccinoh replied:
hmm, this name has been changed before by another reviewer, I think i like `formatTaskList` than the current `printTaskList`, although this gives me problems for the other methods in which similarly, things are not being printed but returned as a string. `formatGreeting` and `formatExitting` both seem like strange names for the other methods following this suggestion...

E0735389 commented on src/main/java/campus/infrastructure/Ui.java:
```diff
@@ -0,0 +1,132 @@
+    public String displayErrorMessage(CampusException e) {
+        String message = "------------------------------\n"
+                + String.format("%s\n", e.getMessage())
+                + "------------------------------\n";
```
Consider factoring out the `--------` to a constant to avoid accidental inconsistency (although looks like you switched to use JavaFX GUI already)

E0735389 commented on src/main/java/campus/javafxGUI/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * An example of a custom control using FXML.
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
```
Consider updating the docstring after you have modified it.

E0735389 commented on src/main/java/campus/tasks/Task.java:
```diff
@@ -0,0 +1,43 @@
+
+    public static void main(String[] args) {
+
+    }
```
What is this method doing here?

E0735389 commented on src/main/java/campus/tasks/Task.java:
```diff
@@ -0,0 +1,43 @@
+
+    }
+
+    public void markComplete() {
```
It would be better to mark the method as abstract so that the subclass does not forget to override it

Chiarena commented on src/main/java/campus/Campus.java:
```diff
@@ -0,0 +1,54 @@
+        return message;
+    }
+
+    public String printExiting() {
```
is there a better way to name this

Kappaccinoh replied:
hmm, im facing this problem else where in other parts of the code in which I return a string after calling the function, it doesnt exactly print but any other naming sounds strange...`returnExitingMessage` seems too long, wdyt?

Chiarena commented on src/main/java/campus/infrastructure/Parser.java:
```diff
@@ -0,0 +1,219 @@
+        this.storage = storage;
+
+        this.taskList = taskList;
+        this.taskList.updateListFromFile(storage.getListOfStrings());
```
comsider using retrieve rather than update for more clarity

Chiarena commented on src/main/java/campus/infrastructure/Parser.java:
```diff
@@ -0,0 +1,219 @@
+        return msg;
+    }
+
+    private String handleTaskCommand(String command, String details) throws CampusException {
```
i like how you named your handle methods

Kappaccinoh replied:
thank you!

Chiarena commented on src/main/java/campus/infrastructure/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+                .collect(Collectors.toList());
+    }
+
+    private Task parseTaskFromStringSafe(String string) {
```
naming is alittle long and not very clear

Kappaccinoh replied:
yea...I think naming could be improved too, would you have any suggestions?

Kappaccinoh committed:
Responding to GH Comments

Several methods have odd naming conventions or have weird names, I'm struggling to come up with a concise name that accurately describes the methods behaviour but doesnt exceed too many characters.

For now I have taken the suggestions on GH but there are still some that are unresolved which I have left open

I also did some refactoring to make the code a little cleaner, and tidied up the abstract class for task

Kappaccinoh committed:
Added Help Page

While the ChatBot has decent features and functionality, a user will be quite clueless as to how to interact with the bot as there is no documentation provided. Even browsing the files doesnt help much since the methods and expectations for commands are spread out across multiple files

I added a simple help command which produces a list of possible commands that the bot can accept

This is to encourage the user to access this help page on the first go for a better user experience with the bot

Kappaccinoh committed:
merge commit with master no ff with C-Help feature branch

Kappaccinoh committed:
Changing code according to code quality conventions feedback

Kappaccinoh committed:
improved GUI

Kappaccinoh committed:
forgot a few files lol

Kappaccinoh committed:
merge commit no ff with betterGUI branch

Kappaccinoh committed:
changed personality

Kappaccinoh committed:
merge commit no ff with personality branch

Kappaccinoh committed:
added AI md file for documentation

Kappaccinoh committed:
merge commit with aiassisted branch

Kappaccinoh committed:
added some junit tests

Kappaccinoh committed:
merge commit with more testing noff

Kappaccinoh committed:
userguide commit

Kappaccinoh committed:
update clean for master

Kappaccinoh committed:
fixing build gradle file

Kappaccinoh committed:
minor edits

Kappaccinoh committed:
final changes to build gradle
